lib/Frontend/Frontend.cpp

```diff
@@ -261,175 +264,281 @@ ModuleDecl *CompilerInstance::getMainModule() {
-                             ImplicitImportModuleName, false);
+        Diagnostics.diagnose(SourceLoc(), diag::sema_no_import,
+                             ImplicitImportModuleName);
+        if (Invocation.getSearchPathOptions().SDKPath.empty() &&
+            llvm::Triple(llvm::sys::getProcessTriple()).isMacOSX()) {
+          Diagnostics.diagnose(SourceLoc(), diag::sema_no_import_no_sdk);
+          Diagnostics.diagnose(SourceLoc(), diag::sema_no_import_no_sdk_xcrun);
+        }
       }
+    } else {
+      Diagnostics.diagnose(SourceLoc(), diag::error_bad_module_name,
+                           ImplicitImportModuleName, false);
     }
   }
+}
 
-  auto addAdditionalInitialImports = [&](SourceFile *SF) {
-    if (!underlying && !importedHeaderModule && importModules.empty())
-      return;
-
-    using ImportPair =
-        std::pair<ModuleDecl::ImportedModule, SourceFile::ImportOptions>;
-    SmallVector<ImportPair, 4> additionalImports;
+void CompilerInstance::createREPLFileWithImports( ImplicitImports &implicitImports) {
+  SharedTimer timer("performSema-createREPLFileWithImports");
+  auto *SingleInputFile =
+      new (*Context) SourceFile(*MainModule, Invocation.getSourceFileKind(),
+                                None, implicitImports.kind);
+  MainModule->addFile(*SingleInputFile);
+  addAdditionalInitialImportsTo(SingleInputFile, implicitImports);
+}
 
-    if (underlying)
-      additionalImports.push_back({ { /*accessPath=*/{}, underlying },
-                                    SourceFile::ImportFlags::Exported });
-    if (importedHeaderModule)
-      additionalImports.push_back({ { /*accessPath=*/{}, importedHeaderModule },
-                                    SourceFile::ImportFlags::Exported });
-    if (!importModules.empty()) {
-      for (auto &importModule : importModules) {
-        additionalImports.push_back({ { /*accessPath=*/{}, importModule }, {} });
-      }
-    }
+DelayedParsingCallbacks *CompilerInstance::computeDelayedParsingCallback() {
+  SharedTimer timer("performSema-computeDelayedParsingCallback");
+  if (Invocation.isCodeCompletion())
+    return new CodeCompleteDelayedCallbacks(SourceMgr.getCodeCompletionLoc());
+  if (Invocation.isDelayedFunctionBodyParsing()) {
+    return new AlwaysDelayedCallbacks;
+  }
+  return nullptr;
+}
 
-    SF->addImports(additionalImports);
-  };
+// The main file may only be
+// a source file, or it may be a SIL file, which requires pumping the parser.
 
-  if (Kind == InputFileKind::IFK_Swift_REPL) {
-    auto *SingleInputFile =
-      new (*Context) SourceFile(*MainModule, Invocation.getSourceFileKind(),
-                                None, modImpKind);
-    MainModule->addFile(*SingleInputFile);
-    addAdditionalInitialImports(SingleInputFile);
+void CompilerInstance::createMainFileWithImportsAndAddToModule(ImplicitImports &implicitImports) {
+  SharedTimer timer("performSema-createMainFileWithImportsAndAddToModule");
+  if (MainBufferID == NO_SUCH_BUFFER)
```

Comment:
It seems like this check belongs outside the function, or perhaps the function should have a "maybe" in its name.

Reply:
Yes, moved the check.